# Day 2 & Lab 1

The goals for today and Lab 1 are: 
* Create a `.gitignore` file
* Be able to import data in both `numpy` and `pandas`
* Articulate the shape of data, specifying 'variables' and 'observations'
* Understand the role of unit testing in our course and in programming development more broadly

### Before starting...

Make sure that you have recently `pull`ed `course-materials`. 

The slides and the labs are closely related. I would recommend following along with the Lab

#### Avoiding Conflicts
The `course-materials` folder is _shared_ but not in the usual `git` sense. 

1. Create a **sub-directory** under `course-materials` called `student-labs`. 
2. Make a copy of the Lab 01 folder.
3. Put the copy in the new sub-directory. 

### `.gitignore`

The `.gitignore` file ignores specific files and folders. 

Check out this file on `github`. 

Following the directions in Lab01, create a `.gitignore` for this lab.

## Creating `.gitignore`

The `.gitignore` file is a [hidden](https://en.wikipedia.org/wiki/Hidden_file_and_hidden_directory) file that tells `git` which files to ignore. In general, you should be creating one for each of your repositories. A few things that you will want to ignore each time include your `.ipynb` checkpoints.    

Refering to the [help file](https://help.github.com/en/articles/ignoring-files), the process for creating `.gitignore` file is as follows: 
1. From your command line, navigate to the directory for your repository on your local machine
2. Next we use the `touch` command (which is similar to `cd` and `m`) to create the file.   
   Type `touch .gitignore`. This creates an empty file with that name
3. Now we need to add the files (or rather the file types) that we do not want to commit to git. To do this, we will use `vim`. Refering to this [Simple Vim Workflow Example](https://www.keycdn.com/blog/vim-commands#simple-vim-workflow-example), we proceed as follows:
   * Type `vim .gitignore`   
     Now you will see an empty file. If you try to type right now, nothing will happen until you hit the **i** key. This is confusing at first. But **i** is a special key in _vim_ as we will discuss next. 
   * Enter **insert mode** by typing `i`. Now you can type whatever you want. Since, we want to ignore our _.ipynb_ checkpoints, add `.ipynb_checkpoints` to your _.gitignore_ file. 
   * Exiting a file in _vim_ is not as simple as closing the window. First, we need to get out of _edit/insert_ mode and back to command mode. To do this hit `Esc`. 
   * Now you are in _command_ mode. To **save and exit** your file, type `:wq` and then `Enter`.  
   
_Unsolicited Advice_ - You will be doing this procedure a lot. I recommend either 1) creating a section in your machine learning notebook for **Repeated Procedures** and adding notes there about how to do this or 2) writing them on the inside cover of your notebook. 

## Importing the necessary packages

The **first coding component** of any script or Jupyter Notebook is the list of imported packages. There are a few reasons for this:
1. **Programming reason** - Python executes in order of the given lines. This means that you need to import a package before you use an element from it
2. **Human reason** - Before running a notebook or a script, a new user would want to know immediately if they can run the script or not. Putting the import statements at the top of the file allows your user to check if they can run or not run your file. 

**Note**: Failure to put your import statments at as the first lines of non-commented code will result in an _automatic loss of half of the assignment's total points_. 

In [1]:
import numpy as np
import pandas as pd

### eRrORs?!?! 

What `conda` enviroment are you in?  

You may need to:
1. Shut down the `jupyter` kernal
2. Activate the correct environment, and 
3. Relaunch your `jupyter notebook`

#### A few notes on imports

You might be wondering why I imported these two packages this way. Partially this is due to what I've seen others do, but a better reason is the one articulated in 'Python for Data Analysis' on page 90:
> ... throughout the book, I use the standard NumPy convention of always using `import numpy as np`. You are, of course, welcome to put `from numpy import *` in your code to avoid having to write `np.`, but I advise against making a habit of this. The `numpy` namespace is large and contains a number of functions whose names conflict with built-in Python functions (like `min` and `max`).

You will even notice that the help files for [numpy](https://docs.scipy.org/doc/numpy/reference/arrays.ndarray.html) and [pandas](https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#basics) use these conventions. 

## Importing Data

When working with data, we need to first import that data. 

One can do this either using `pandas` or using `numpy`. 

Today, we will work with both methods for sake of completeness. 

### Importing with `numpy`

Importing with `numpy` will bring in your data as a `numpy array`. The most straightforward way to do this is using `genfromtxt()`. To learn more about this function, read its [help file](https://docs.scipy.org/doc/numpy/reference/generated/numpy.genfromtxt.html). 

In [2]:
ffire_np = np.genfromtxt("forestfires.csv", delimiter=',')

Taking this apart piece by piece, let's look at what we just did:
1. `ffire_np` is a variable
2. `np.` tells us to reach into the `numpy` library
3. `genfromtxt` is the specific method that we want to call
4. The first argument is the name of the file, which we have placed within the same directory. 
5. The second argument `delimiter=` tells us what gaps to look for between data information. It is safe to use a comma as our separation since `csv` stands for `c`omma `s`eparated `v`aribles. 

Let's take a peek at what our variable looks like:

In [3]:
print(ffire_np)

[[  nan   nan   nan ...   nan   nan   nan]
 [ 7.    5.     nan ...  6.7   0.    0.  ]
 [ 7.    4.     nan ...  0.9   0.    0.  ]
 ...
 [ 7.    4.     nan ...  6.7   0.   11.16]
 [ 1.    4.     nan ...  4.    0.    0.  ]
 [ 6.    3.     nan ...  4.5   0.    0.  ]]


Is this what we expect to see? Open the datafile using your favorite spreadsheet viewer. 

What do you see? Does it match the above? 

### More arguments

There is a third argument that we can use with `genfromtxt` to ignore the column headers: `skip_header=1`. 

Use `genfromtxt` with three arguments to re-import the `forestfires` data and print the result. 

In [ ]:
# Import forestfire data here
ffire_np = 

# Take a look at the result here




Again, is this what you expect to see? Compare this with the open datafile in your favorite spreadsheet viewer. 

#### Limitations in `numpy`

`numpy` can have a hard time with non-numerical data. When `numpy` doesn't know what to do with cells that are not numbers, it replaces those cells with `nan` or `n`ot `a` `n`umber. 

Take a look at both of your outputs, and comparing them again to the open datafile in your favorite spreadsheet viewer. Which rows and colums have `nan`s? 

#### What? Why?

This doesn't feel particularly useful given that so much data contains information that is non-numerical. However, a closer examination of what `numpy` stands for makes it a bit clearer why `numpy` does this. With each reference to `numpy`, I've used the `code` formatting, but the name of this package is **NumPy** or **Num**erical **Py**thon. This package is for the fast processing of numerical data leveraging tricks from linear algebra. 

`pandas` can offer us a bit more flexibility. Let's try importing data with it. 

### Importing with `pandas`

`pandas` has more functionality when it comes to data with non-numerical values than `numpy`. While our book doesn't delve into the root of the name `panda`, according to [this site](https://towardsdatascience.com/a-quick-introduction-to-the-pandas-python-library-f1b678f34673), `pandas` is short for `P`ython `Da`ta `An`aly`s`is. (Clearly I'm not sure how they got 'pandas' from those three words, so if you would like to instead just imagine a group of cuddly pandas, that's fine with me.) 

Returning to how we can import using `pandas`, the most common method that we will use is `read_csv()`.  To learn more about this function, read its [help file](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.read_csv.html). 

In [4]:
ffire_pd = pd.read_csv("forestfires.csv", sep=',')

Taking this apart piece by piece, let's look at what we just did:
1. `ffire_pd` is again just a variable
2. `pd.` tells us to reach into the `pandas` library
3. `read_csv` is the specific method that we want to call
4. The first argument is the name of the file (just like in `genfromtxt()` from above).   
5. The second argument `sep=` is the pandas version of `delimiter=` telling us what gaps to look for between data information. (Why are we using a comma here for denoting the gaps?)

Let's take a peek at what our variable looks like. Note that we're going to do this both using and not using the `print()` method. 

In [5]:
ffire_pd

X  Y month  day  FFMC    DMC     DC   ISI  temp  RH  wind  rain   area
0    7  5   mar  fri  86.2   26.2   94.3   5.1   8.2  51   6.7   0.0   0.00
1    7  4   oct  tue  90.6   35.4  669.1   6.7  18.0  33   0.9   0.0   0.00
2    7  4   oct  sat  90.6   43.7  686.9   6.7  14.6  33   1.3   0.0   0.00
3    8  6   mar  fri  91.7   33.3   77.5   9.0   8.3  97   4.0   0.2   0.00
4    8  6   mar  sun  89.3   51.3  102.2   9.6  11.4  99   1.8   0.0   0.00
5    8  6   aug  sun  92.3   85.3  488.0  14.7  22.2  29   5.4   0.0   0.00
6    8  6   aug  mon  92.3   88.9  495.6   8.5  24.1  27   3.1   0.0   0.00
7    8  6   aug  mon  91.5  145.4  608.2  10.7   8.0  86   2.2   0.0   0.00
8    8  6   sep  tue  91.0  129.5  692.6   7.0  13.1  63   5.4   0.0   0.00
9    7  5   sep  sat  92.5   88.0  698.6   7.1  22.8  40   4.0   0.0   0.00
10   7  5   sep  sat  92.5   88.0  698.6   7.1  17.8  51   7.2   0.0   0.00
11   7  5   sep  sat  92.8   73.2  713.0  22.6  19.3  38   4.0   0.0   0.00
12   6  5   aug  fri  63.5   70.8  665.3   0.8  17.0  72   6.7   0.0   0.00
13   6  5   sep  mon  90.9  126.5  686.5   7.0  21.3  42   2.2   0.0   0.00
14   6  5   sep  wed  92.9  133.3  699.6   9.2  26.4  21   4.5   0.0   0.00
15   6  5   sep  fri  93.3  141.2  713.9  13.9  22.9  44   5.4   0.0   0.00
16   5  5   mar  sat  91.7   35.8   80.8   7.8  15.1  27   5.4   0.0   0.00
17   8  5   oct  mon  84.9   32.8  664.2   3.0  16.7  47   4.9   0.0   0.00
18   6  4   mar  wed  89.2   27.9   70.8   6.3  15.9  35   4.0   0.0   0.00
19   6  4   apr  sat  86.3   27.4   97.1   5.1   9.3  44   4.5   0.0   0.00
20   6  4   sep  tue  91.0  129.5  692.6   7.0  18.3  40   2.7   0.0   0.00
21   5  4   sep  mon  91.8   78.5  724.3   9.2  19.1  38   2.7   0.0   0.00
22   7  4   jun  sun  94.3   96.3  200.0  56.1  21.0  44   4.5   0.0   0.00
23   7  4   aug  sat  90.2  110.9  537.4   6.2  19.5  43   5.8   0.0   0.00
24   7  4   aug  sat  93.5  139.4  594.2  20.3  23.7  32   5.8   0.0   0.00
25   7  4   aug  sun  91.4  142.4  601.4  10.6  16.3  60   5.4   0.0   0.00
26   7  4   sep  fri  92.4  117.9  668.0  12.2  19.0  34   5.8   0.0   0.00
27   7  4   sep  mon  90.9  126.5  686.5   7.0  19.4  48   1.3   0.0   0.00
28   6  3   sep  sat  93.4  145.4  721.4   8.1  30.2  24   2.7   0.0   0.00
29   6  3   sep  sun  93.5  149.3  728.6   8.1  22.8  39   3.6   0.0   0.00
..  .. ..   ...  ...   ...    ...    ...   ...   ...  ..   ...   ...    ...
487  5  4   aug  tue  95.1  141.3  605.8  17.7  26.4  34   3.6   0.0  16.40
488  4  4   aug  tue  95.1  141.3  605.8  17.7  19.4  71   7.6   0.0  46.70
489  4  4   aug  wed  95.1  141.3  605.8  17.7  20.6  58   1.3   0.0   0.00
490  4  4   aug  wed  95.1  141.3  605.8  17.7  28.7  33   4.0   0.0   0.00
491  4  4   aug  thu  95.8  152.0  624.1  13.8  32.4  21   4.5   0.0   0.00
492  1  3   aug  fri  95.9  158.0  633.6  11.3  32.4  27   2.2   0.0   0.00
493  1  3   aug  fri  95.9  158.0  633.6  11.3  27.5  29   4.5   0.0  43.32
494  6  6   aug  sat  96.0  164.0  643.0  14.0  30.8  30   4.9   0.0   8.59
495  6  6   aug  mon  96.2  175.5  661.8  16.8  23.9  42   2.2   0.0   0.00
496  4  5   aug  mon  96.2  175.5  661.8  16.8  32.6  26   3.1   0.0   2.77
497  3  4   aug  tue  96.1  181.1  671.2  14.3  32.3  27   2.2   0.0  14.68
498  6  5   aug  tue  96.1  181.1  671.2  14.3  33.3  26   2.7   0.0  40.54
499  7  5   aug  tue  96.1  181.1  671.2  14.3  27.3  63   4.9   6.4  10.82
500  8  6   aug  tue  96.1  181.1  671.2  14.3  21.6  65   4.9   0.8   0.00
501  7  5   aug  tue  96.1  181.1  671.2  14.3  21.6  65   4.9   0.8   0.00
502  4  4   aug  tue  96.1  181.1  671.2  14.3  20.7  69   4.9   0.4   0.00
503  2  4   aug  wed  94.5  139.4  689.1  20.0  29.2  30   4.9   0.0   1.95
504  4  3   aug  wed  94.5  139.4  689.1  20.0  28.9  29   4.9   0.0  49.59
505  1  2   aug  thu  91.0  163.2  744.4  10.1  26.7  35   1.8   0.0   5.80
506  1  2   aug  fri  91.0  166.9  752.6   7.1  18.5  73   8.5   0.0   0.00
507  2  4   aug  fri  91.0  166.9  752.6   7.1  25.9 

In [6]:
print(ffire_pd)

     X  Y month  day  FFMC    DMC     DC   ISI  temp  RH  wind  rain   area
0    7  5   mar  fri  86.2   26.2   94.3   5.1   8.2  51   6.7   0.0   0.00
1    7  4   oct  tue  90.6   35.4  669.1   6.7  18.0  33   0.9   0.0   0.00
2    7  4   oct  sat  90.6   43.7  686.9   6.7  14.6  33   1.3   0.0   0.00
3    8  6   mar  fri  91.7   33.3   77.5   9.0   8.3  97   4.0   0.2   0.00
4    8  6   mar  sun  89.3   51.3  102.2   9.6  11.4  99   1.8   0.0   0.00
5    8  6   aug  sun  92.3   85.3  488.0  14.7  22.2  29   5.4   0.0   0.00
6    8  6   aug  mon  92.3   88.9  495.6   8.5  24.1  27   3.1   0.0   0.00
7    8  6   aug  mon  91.5  145.4  608.2  10.7   8.0  86   2.2   0.0   0.00
8    8  6   sep  tue  91.0  129.5  692.6   7.0  13.1  63   5.4   0.0   0.00
9    7  5   sep  sat  92.5   88.0  698.6   7.1  22.8  40   4.0   0.0   0.00
10   7  5   sep  sat  92.5   88.0  698.6   7.1  17.8  51   7.2   0.0   0.00
11   7  5   sep  sat  92.8   73.2  713.0  22.6  19.3  38   4.0   0.0   0.00
12   6  5   

#### Before we go on:
1. What do you see in both of these views of the output `ffire_pd`? 
2. What are the differences between these views?
3. How does `ffire_pd` compare to the datafile in your favorite spreadsheet viewer?
4. How do `ffire_pd` and `ffire_np` differ? 

### Data Frames and Series

Data types are something that we pay a lot of attention to in computer science. So before moving much further,we should ask ourselves: what kinds of objects did we just create? (How do we check the **type** of an object in python?)

In [ ]:
# Print the data types for ffire_np and ffire_pd 




We'll turn our attention to the `pandas` version: `ffire_pd`. 

This variable is a `pandas` **_DataFrame_**, which are comprised of **_Series_**. 
* DataFrames look like spreadsheets that we are used to seeing, containing both numerical and non-numerical values
* Each row of the DataFrame is a Series.

## Shape of Data

Data, like spreadsheets, have a shape. Most data that we will work with in this class will have a shape that can be described by the number of rows and the number of columns. When refering to data, we think of **_Observations_** and _**Variables**_. 
* Each observation is a row. Think of the observations as an object, person, or item that we have a set of information on. 
* The variables are stored as columns. The variables details the kind of information that we have stored on each observation. 

Looking at your dataframe and refering to the data's [code book](http://archive.ics.uci.edu/ml/datasets/Forest+Fires), what are the observations in this dataset? What are the variables? _(Note: A code book is a document that provides details on a dataset.)_

#### Finding the shape of your data

Once you have the data loaded, you can quickly generate a number of facts about your data:
1. The dimensions of your data (i.e. the number of observations and variables) using `.shape`
2. The names of the variables using `.columns`
3. Create a quick glance of your data using `.head`

In [8]:
# 1. Using `.shape` on ffire_pd
print(ffire_pd.shape)

(517, 13)


In [9]:
#2. Using `.columns` on ffire_pd
var_names = ffire_pd.columns
print(var_names)

Index(['X', 'Y', 'month', 'day', 'FFMC', 'DMC', 'DC', 'ISI', 'temp', 'RH',
       'wind', 'rain', 'area'],
      dtype='object')


In [10]:
# 3. Using `.head` on ffire_pd
glance = ffire_pd.head()
print(glance)

   X  Y month  day  FFMC   DMC     DC  ISI  temp  RH  wind  rain  area
0  7  5   mar  fri  86.2  26.2   94.3  5.1   8.2  51   6.7   0.0   0.0
1  7  4   oct  tue  90.6  35.4  669.1  6.7  18.0  33   0.9   0.0   0.0
2  7  4   oct  sat  90.6  43.7  686.9  6.7  14.6  33   1.3   0.0   0.0
3  8  6   mar  fri  91.7  33.3   77.5  9.0   8.3  97   4.0   0.2   0.0
4  8  6   mar  sun  89.3  51.3  102.2  9.6  11.4  99   1.8   0.0   0.0


Taken these pieces apart: 
1. `.shape` tells us the number of observations followed by the number of variables. We call data _tall_ if there are substantially more observations than variables. We say data is _wide_ if there are substantially more variables than observations. 
2. `.columns` gives us the list of _variables_ which `pandas` considers as an index. 
3. `.head` gives us a snippet of the data. The default is 5 rows, but we can actually specify the number of rows that we want to see by supplying a number between the parentheses. 

#### Your Turn

Check out [fivethirtyeight's data repository on GitHub](https://github.com/fivethirtyeight/data) and pick a dataset of interest. Use python to import the data and describe key features about the data set. Practice your machine learning communication skills and write a short paragraph about your data based on the code that you ran. 

In [ ]:
# Code Block for you!




#### (Markdown block for you!)






## Powerhouse packages


## Powerhouse packages

As `python` and data science have grown and evolved, several packages -- including `numpy`, `pandas`, `scipy`, `matplotlib` -- as critical to the practice of machine learning. In this course, we will use these packages, but with an eye towards deep understand of each method that we employ. 

This course has twin themes: **carpentry** and **creativity**. Most of the assignments in the course focus on the former in service of the latter. Think of this course as a kind of cookng class where we will spend a considerable amount of time on each ingredient and the most basic of recipes, with the ultimate goal of creating a glorious meal mixing and blending the ingredients in unexpected ways with each other and with new ingredients. 

**Carpentry**: We are going to focus on the interior elements of classic machine learning algorithms, building each one from scratch. We will compare our results to already optimized versions in packages like `scikit-learn`. The goal of carpentry is _deep_ understanding of each algorithm. The course _**Homework**_ and _**Labs**_ are designed to build your toolkit. 

**Creativity**: While the majority of class time and homework assignments are focused on carpentry, we are building towards you asking "what if I applied X to Y in this way?" and "what happens if we remove this assumption from Z?" The goal of creativity is applying these algorithms to problems and data that you care about in _new_ and _exciting_ ways. The course **_Projects_** are designed to engage you creatively. 

## Unit Testing: When are we right? 

We will be using unit testing in our course a lot. In fact, each homework assignment will have unit tests that you can use (or not). Quoting [McFee and Kell](https://bmcfee.github.io/ismir2018-oss-tutorial/tutorial/2018/08/12/part-5.html): 

> Tests make your life easier and give you confidence in your code. Software is complicated, and tests allow you to be sure that you’ve not broken your project when you change something.
>
>Writing tests slows you down in the short term, but speeds things up in the long term. Further, as someone who is writing scientific code, tests make sure that your work can be easily reproduced.
>
>Philosophically, each test should test one thing. This makes it easier to ensure that we can see what has broken when a test breaks. 

In the next few slides, we will go over:
1. Using tests
2. Writing tests

#### Unit tests in our course

For many of our course's homework assignments, you will be given tests. You can use these tests to check your code and see if it is working how I expect it to. In the case of projects, you may be asked to design your own tests. Even if you are not explicitly asked to design a test, it is strongly recommended that you get in the habit of building tests. 

A few notes: 
1. When grading your homework, The first thing I will do is check to see if the tests pass. 
2. I personally am just getting in the habit of writing and using unit-tests. 

### Using tests

We will use `pytest` to test our work. In this course, the convention for these files will to have a file name that begins with `test`. 

According to [this post](https://semaphoreci.com/community/tutorials/testing-python-applications-with-pytest), `pytest` expects our test files to either use this preamble or end with `_test.py`. In this section, I am again editing from [Part 5](https://bmcfee.github.io/ismir2018-oss-tutorial/tutorial/2018/08/12/part-5.html) of McFee and Kell's tutorial: 

To run the tests, from your command line, type:
`pytest -v`

Note:
1. My test files will assume that you have followed all directions in the homework assignment, including and especially naming files correctly. Please pay careful attention to these details. 
2. You cannot test functions inside a `jupyter notebook`. You will need to create a `.py` file. 

### Your turn

I have written a set of 3 tests for this lab in a file called `test_Lab1.py`. Open this file in your favorite code editor and then run the tests from the command line. What happens?   

They failed?! Oh no! <sub>This was intentional. Tests fail a lot. It's ok.</sub>   

Let's figure out what is going on. (Remember that **I** wrote these tests. You have done nothing wrong.)   

Read the error message that you got. What seems to be the issue?    

Test your theory by opening `labone.py`. Make **2** minor changes to this file and run the tests again. (When they pass, take a moment to celebrate. Happy dances are encouraged.) 

#### Note about imports in testing

You will notice that we have three odd lines in the test file:
1. `import os`
2. `path, _ = os.path.split(os.path.abspath(__file__))`
3. `fname = os.path.join(path,"forestfires.csv")`

When testing our files, Travis is building a virtual machine to test our files. We need the paths to be such that Travis can build them. The first line imports the `os` package that allows us to do path manipulations. The second line extracts the path for the test file and splits it such that the file name (in this case `test_Lab1.py`) from the directory name (i.e. where the test file lives). Then the third line connects our directory with the data file. 

### Writing tests

To write your own tests, you need to 
1. Create a python file whose name starts with `test`
2. `import` the python file that you are testing 
3. Write out what you expect to happen
4. Write an assert statement to test the results of your code against your expectations

#### Your turn

Edit `test_importdata` to test the function `importdata` on importing your dataset from earlier. 

### Final Thoughts

To finish up this lab, create a post to **#lab_submission** channel on slack sharing the green box of all your tests passing on travis. Your post must start with **Lab1** to get credit. 

If your have questions from this lab, post them to #lab_questions with the same preamble (i.e. starting with **Lab1**). If you have the same question, please use one of the emoji's to upvote the question. If you would like to answer someone's question, please use the thread function. This will tie your answer to their question. 

#### Resources consulted to build these slides and associated lab:

1. _Python for Data Analysis_ 
2. [Open Source and Reproducible MIR Research](https://bmcfee.github.io/ismir2018-oss-tutorial/)
3. [NumPy Tutorial: Data analysis with Python](https://www.dataquest.io/blog/numpy-tutorial-python/) on Dataquest
4. [A Quick Introduction to the “Pandas” Python Library](https://towardsdatascience.com/a-quick-introduction-to-the-pandas-python-library-f1b678f34673)
5. [Testing Python Applications with Pytest](https://semaphoreci.com/community/tutorials/testing-python-applications-with-pytest)
6. [Find current directory and file's directory [duplicate]](https://stackoverflow.com/questions/5137497/find-current-directory-and-files-directory)
7. [Travis CI tests (Python 3.5, 3.6) failing on reproducibility tests](https://github.com/soft-matter/trackpy/issues/459)
8. [Ignoring Files and Directories in Git with .gitignore](https://www.jamescoyle.net/how-to/1094-ignoring-files-in-git-with-gitignore)
9. [ignoring any 'bin' directory on a git project](https://stackoverflow.com/questions/1470572/ignoring-any-bin-directory-on-a-git-project)